### DISCUSSION TITLE
In NMS, what does the "offset by class" do ?

### DISCUSSION THREAD

**Discussion by gboeer** [ultralytics/yolov5#5825] [Q&A] [answered]
In the non maximum suppression, I was wondering what this line of code does. Maybe someone can give me a hint. I am asking specifically, because I want to turn on the agnostic behavior, but only for some of my classes, not all. I guess I can add it here somehow but struggle to understand the reason for this "class offset".

https://github.com/ultralytics/yolov5/blob/5ca5dd4c87fcc62491173b393fd51cf244805313/utils/general.py#L713-L714

**Comment by nrupatunga** [comment, marked as answer]
Here is my understanding. 

If you check the torchvision NMS function
`i = torchvision.ops.nms(boxes, scores, iou_thres) ` it does not take class information as input and when `agnostic=False`, we need to use the same function somehow, so adding class offset is a small trick that helps to move the overlapping boxes of different classes to different positions, which will help in performing the class-wise NMS using the same function. 

The figure below illustrates the same. Red and green boxes are for different classes with offset `c1, c2` . 
Adding offset shifts the boxes accordingly that help in doing NMS using `torchvision.ops.nms(boxes, scores, iou_thres)`

![image](https://user-images.githubusercontent.com/980580/143913745-2e657530-0ace-4a90-80a7-36d67833f9c7.png)

**Reply by gboeer** [reply]
Hi @nrupatunga 
thanks for taking the time to illustrate this, it really helped and I get it now. Its indeed a small trick to spatially separate the boxes of the same class by using the respective class_id and multiply it with a offset (the maximum allowed box size).

![image](https://user-images.githubusercontent.com/1067159/144016525-71cff156-7c89-497b-ab4c-09e5a2f71c7f.png)

**Comment by beneon** [comment]
but doesn't `torchvision.ops.boxes.batched_nms` receive label? So why using this trick?
